include/private/net/libssl.hpp

```diff
@@ -0,0 +1,331 @@
+    ~Context() {
+        assert(ctx_ != nullptr);
+        SSL_CTX_free(ctx_);
+    }
+
+    /*!
+        \brief Factory that creates a shared `SSL_CTX *` wrapper.
+
+        \param ca_bundle_path Path to the CA bundle to be used to verify
+        the servers we're connecting to. If it is empty and we're linking
+        with libressl, we will use the builtin CA bundle that is shipped
+        along with measurement-kit. Otherwise, we will fail.
+
+        \param logger Logger for printing log messages.
+
+        \return An error on failure (the returned object will evaluate to
+        false in a boolean context) or a `Var` shared pointer wrapping a
+        Context instance on success.
+    */
+    template <MK_MOCK(SSL_CTX_new), MK_MOCK(SSL_CTX_load_verify_locations)
+#if (defined LIBRESSL_VERSION_NUMBER && LIBRESSL_VERSION_NUMBER >= 0x2010400fL)
+                                          ,
+              MK_MOCK(SSL_CTX_load_verify_mem)
+#endif
+              >
+    static ErrorOr<Var<Context>> make(std::string ca_bundle_path,
+                                      Var<Logger> logger) {
+        // Implementation note: we need to get the library singleton as early
+        // as here to make sure OpenSSL internals are initialized.
+        auto library = Library::global();
+        logger->debug("ssl: creating SSL_CTX with bundle: '%s'",
+                      ca_bundle_path.c_str());
+        SSL_CTX *ctx = SSL_CTX_new(SSLv23_client_method());
+        if (ctx == nullptr) {
+            logger->warn("ssl: failed to create SSL_CTX");
+            return SslCtxNewError();
+        }
+        /*
+         * Implementation note:
+         *
+         * We use SSLv23_client_method() because it is the most general method
+         * that applications SHOULD use according to the manual[1] and we
+         * disable by default deprecated protocols. We MAY want to enable them
+         * in the OONI's web_connectivity test so to reach more websites.
+         *
+         * [1] https://wiki.openssl.org/index.php/Manual:SSL_CTX_new(3)
+         */
+        SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+        if (ca_bundle_path != "") {
+            if (!SSL_CTX_load_verify_locations(ctx, ca_bundle_path.c_str(),
+                                               nullptr)) {
+                logger->warn("ssl: failed to load verify location");
+                SSL_CTX_free(ctx);
+                return SslCtxLoadVerifyLocationsError();
+            }
+        } else {
+#if (defined LIBRESSL_VERSION_NUMBER && LIBRESSL_VERSION_NUMBER >= 0x2010400fL)
+            std::vector<uint8_t> bundle = builtin_ca_bundle();
+            logger->debug("ssl: using builtin libressl's ca bundle");
+            if (!SSL_CTX_load_verify_mem(ctx, bundle.data(), bundle.size())) {
+                logger->warn("ssl: failed to load default ca bundle");
+                SSL_CTX_free(ctx);
+                return SslCtxLoadVerifyMemError();
+            }
```

Comment:
Note: we currently don't have coverage of this section. To have it, we need to link against libressl. (Or, has OpenSSL picked up this libressl function? Or, can I write a diff for OpenSSL that ports this function?)

Reply:
For now, I guess we can ignore this problem.